File: shoring/flywheel/src/main/java/com/panopset/flywheel/TemplateSource.java
package com.panopset.flywheel;

public interface TemplateSource {
  boolean isDone();
  String nextRow();
  default String next() {
    String rtn = nextRow();
    boolean lastBackslashIsReal = false;
    if (rtn.length() > 0) {
      if (rtn.length() > 1) {
        String last2 = rtn.substring(rtn.length() - 2);
        if ("\\\\".equals(last2)) {
          lastBackslashIsReal = true;
        }
      }
      rtn = rtn.replaceAll("\\\\\\\\", "\\\\");
      if (!lastBackslashIsReal) {
        String lastChar = rtn.substring(rtn.length() - 1);
        if (!lastBackslashIsReal && "\\".equals(lastChar) && !isDone()) {
          String s = rtn.substring(0, rtn.length() - 1);
          if (isDone()) {
            return s;
          } else {
            return String.format("%s%s", s, next());
          }
        }
      }
    }
    return rtn;
  }
  void reset();
  int getLine();
  String getName();
  String getRaw();
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/TemplateSource.kt
package com.panopset.flywheel

interface TemplateSource {
    val isDone: Boolean
    fun nextRow(): String
    fun nextTemplateSourceLine(): String {
        var rtn = nextRow()
        var lastBackslashIsReal = false
        if (rtn.isNotEmpty()) {
            if (rtn.length > 1) {
                val last2 = rtn.substring(rtn.length - 2)
                if ("\\\\" == last2) {
                    lastBackslashIsReal = true
                }
            }
            rtn = rtn.replace("\\\\\\\\".toRegex(), "\\\\")
            if (!lastBackslashIsReal) {
                val lastChar = rtn.substring(rtn.length - 1)
                if ("\\" == lastChar && !isDone) {
                    val s = rtn.substring(0, rtn.length - 1)
                    return if (isDone) {
                        s
                    } else {
                        String.format("%s%s", s, nextTemplateSourceLine())
                    }
                }
            }
        }
        return rtn
    }

    fun reset()
    val line: Int
    val name: String?
    val raw: String?
}
